library/bignum.c

```diff
@@ -258,6 +258,10 @@ static inline mbedtls_mpi_uint mpi_sint_abs(mbedtls_mpi_sint z)
     return (mbedtls_mpi_uint) 0 - (mbedtls_mpi_uint) z;
 }
 
+/* Convert x to a sign, i.e. to 1, if x is positive, or -1, if x is negative.
+ * This looks awkward but generates smaller code than (x < 0 ? -1 : 1) */
+#define TO_SIGN(x) ((((mbedtls_mpi_uint) x) >> (biL - 1)) * -2 + 1)
```

Comment:
How did you work out this optimisation?

Reply:
I suspected the conditional might generate several instructions - turns out, it generates this:

```
   0:	2101      	movs	r1, #1
   2:	2801      	cmp	r0, #1
   4:	bfb8      	it	lt
   6:	f04f 31ff 	movlt.w	r1, #4294967295	; 0xffffffff
   a:	4608      	mov	r0, r1
```

and it looked like something that could possibly be done with a bit of bit-twiddling - expressing it this way enables the compiler to find:

```
   0:	2101      	         movs	r1, #1
   2:	ea41 70a0 	orr.w	r0, r1, r0, asr #30
```

I guess it could be written as `#define TO_SIGN(x) (((x) >> 30)|1)` but the version I pushed seemed more human-readable.

Reply:
Yes, I looked at the disassembly of the `.o` files and saw the same thing - there's an extra register save in the bitwise case, but it's just an extra bit set in the `push` in the function prologue. Was curious as to how you found it, wondering if it was a technique we could use to find other improvements.

I should actually check what it does to code size on other platforms!

Reply:
I think basically, the key observation was noticing conditional code that could be replaced with bit-twiddling. Then trial and error with 'mtest -d'.